## **Tratamento do Conjunto de Dados** 

In [15]:
import pandas as pd

> Carregando o conjunto de dados como um DataFrame Pandas.

In [16]:
df = pd.read_csv('data\Dados-Abertos.csv', delimiter=';', encoding='latin-1')

In [17]:
df.head()

Nome Ente  UF Código Siafi  Código IBGE   Data   Ano      Mês  \
0  Nova Santa Rosa  PR         7979      4117222  42005  2015  janeiro   
1          Humaitá  RS         8695      4309704  42005  2015  janeiro   
2     Nova Venécia  ES         5677      3203908  42005  2015  janeiro   
3         Gravatal  SC         8121      4206207  42005  2015  janeiro   
4  Passo de Torres  SC         5541      4212254  42005  2015  janeiro   

   Tipo Ente                       OB  CNPJ do Favorecido  \
0  Município  135098000012015OB800001      77116663000109   
1  Município  135098000012015OB800002      87613139000199   
2  Município  135098000012015OB800003      27167428000180   
3  Município  135098000012015OB800004      82926569000147   
4  Município  135098000012015OB800005      95782793000154   

                Nome Favorecido        Nome Emenda Transferência Especial  \
0  MUNICIPIO DE NOVA SANTA ROSA  Emenda Individual                    Não   
1          MUNICIPIO DE HUMAITA  Emenda Individual                    Não   
2     MUNICIPIO DE NOVA VENECIA  Emenda Individual                    Não   
3         MUNICIPIO DE GRAVATAL  Emenda Individual                    Não   
4  MUNICIPIO DE PASSO DE TORRES  Emenda Individual                    Não   

  Categoria Econômica Despesa   Valor  
0         DESPESAS DE CAPITAL  243750  
1         DESPESAS DE CAPITAL  144300  
2         DESPESAS DE CAPITAL   97500  
3         DESPESAS DE CAPITAL  195000  
4         DESPESAS DE CAPITAL   97500

In [18]:
df.dtypes

Nome Ente                      object
UF                             object
Código Siafi                   object
Código IBGE                     int64
Data                            int64
Ano                             int64
Mês                            object
Tipo Ente                      object
OB                             object
CNPJ do Favorecido              int64
Nome Favorecido                object
Nome Emenda                    object
Transferência Especial         object
Categoria Econômica Despesa    object
Valor                          object
dtype: object

In [19]:
example = df.head(1).copy()
example

Nome Ente  UF Código Siafi  Código IBGE   Data   Ano      Mês  \
0  Nova Santa Rosa  PR         7979      4117222  42005  2015  janeiro   

   Tipo Ente                       OB  CNPJ do Favorecido  \
0  Município  135098000012015OB800001      77116663000109   

                Nome Favorecido        Nome Emenda Transferência Especial  \
0  MUNICIPIO DE NOVA SANTA ROSA  Emenda Individual                    Não   

  Categoria Econômica Despesa   Valor  
0         DESPESAS DE CAPITAL  243750

> Checando se existem valores acima da casa dos milhões para a coluna 'Valor'.

In [20]:
million_unit = 7
(df['Valor'].apply(len) > million_unit).any()

True

> Checando qual o tamanho máximo da string para 'Valor'.

In [21]:
df['Comprimento'] = df['Valor'].apply(len)
df['Comprimento'].max()

11

> Obtendo valores de exemplo para os tamanhos.

In [22]:
max_length = 11
filtered = df[df['Comprimento'] == max_length]
filtered.head()

Nome Ente  UF Código Siafi  Código IBGE   Data   Ano       Mês  \
50025         Pernambuco  PE           PE           26  43070  2017  dezembro   
50026         Pernambuco  PE           PE           26  43070  2017  dezembro   
50921              Bahia  BA           BA           29  43160  2018     março   
50932  Rio Grande do Sul  RS           RS           43  43160  2018     março   
61972              Goiás  GO           GO           52  43252  2018     junho   

      Tipo Ente                       OB  CNPJ do Favorecido  \
50025    Estado  530013000012017OB801718      10571982000125   
50026    Estado  530013000012017OB801848      10571982000125   
50921    Estado  135098000012018OB800139      13937057000163   
50932    Estado  135098000012018OB800154      93021632000112   
61972    Estado  135098000012018OB800672      21652711000110   

                                         Nome Favorecido        Nome Emenda  \
50025                               ESTADO DE PERNAMBUCO  Emenda de Bancada   
50026                               ESTADO DE PERNAMBUCO  Emenda de Bancada   
50921  SECRETARIA DA AGRICULTURA, PECUARIA, IRRIGACAO...  Emenda de Bancada   
50932  SECRETARIA DE AGRICULTURA, PECUARIA E DESENVOL...  Emenda de Bancada   
61972  SECRETARIA DE ESTADO DE DESENVOLVIMENTO E INOV...  Emenda de Bancada   

      Transferência Especial Categoria Econômica Despesa        Valor  \
50025                    Não         DESPESAS DE CAPITAL  68217149,52   
50026                    Não         DESPESAS DE CAPITAL  58132127,48   
50921                    Não         DESPESAS DE CAPITAL  61575657,97   
50932                    Não         DESPESAS DE CAPITAL  18779613,45   
61972                    Não         DESPESAS DE CAPITAL  82323934,88   

       Comprimento  
50025           11  
50026           11  
50921           11  
50932           11  
61972           11

> Observou-se uma inconsistência nos dados, na qual alguns valores apresentam precisão até os centavos, enquanto outros não mantêm a mesma precisão decimal para valores na mesma casa decimal.
- 68217149,52 (tamanho 11)
- 1470678,32 (tamanho 10)
- 872142,22 (tamanho 9)
- 203528,5 (tamanho 8)
- 1358175 (tamanho 7)
- 89035,5 (tamanho 7)

## **Geração de Sentenças**

In [23]:
def generate_sentence(data):
    sentence = (
        f"Em {data['Mês']} de {data['Ano']}, o {data['Tipo Ente']} de {data['Nome Ente']} - {data['UF']}, "
        f"com o código Siafi {data['Código Siafi']} e o código IBGE {data['Código IBGE']}, "
        f"recebeu uma transferência no valor de R$ {data['Valor']}. A {data['Nome Emenda']} "
        f"parlamentar classificada como {data['Categoria Econômica Despesa']} teve como favorecido "
        f"o CNPJ {data['CNPJ do Favorecido']}, identificado pelo nome {data['Nome Favorecido']} "
        f"com a ordem bancária {data['OB']}."
    )
    return sentence

In [24]:
df.head(1)

Nome Ente  UF Código Siafi  Código IBGE   Data   Ano      Mês  \
0  Nova Santa Rosa  PR         7979      4117222  42005  2015  janeiro   

   Tipo Ente                       OB  CNPJ do Favorecido  \
0  Município  135098000012015OB800001      77116663000109   

                Nome Favorecido        Nome Emenda Transferência Especial  \
0  MUNICIPIO DE NOVA SANTA ROSA  Emenda Individual                    Não   

  Categoria Econômica Despesa   Valor  Comprimento  
0         DESPESAS DE CAPITAL  243750            6

"Em **janeiro** de **2015**, o **Município** de **Nova Santa Rosa** - **PR**, com o código Siafi **7979** e o código IBGE **4117222**, recebeu uma transferência no valor de R$ **243750**. A **Emenda Individual** parlamentar classificada como **DESPESAS DE CAPITAL** teve como favorecido o CNPJ **77116663000109**, identificado pelo nome **MUNICIPIO DE NOVA SANTA ROSA** com a ordem bancária **135098000012015OB800001**."


In [25]:
print(generate_sentence(df.iloc[0]))

Em janeiro de 2015, o Município de Nova Santa Rosa - PR, com o código Siafi 7979 e o código IBGE 4117222, recebeu uma transferência no valor de R$ 243750. A Emenda Individual parlamentar classificada como DESPESAS DE CAPITAL teve como favorecido o CNPJ 77116663000109, identificado pelo nome MUNICIPIO DE NOVA SANTA ROSA com a ordem bancária 135098000012015OB800001.


In [26]:
sentences = df.apply(generate_sentence, axis=1).to_list()

## **Tokenização e Geração de Embeddings**

In [27]:
df.shape

(295473, 16)

> O conjunto de dados original possui 295.473 registros, contudo, serão utilizados somente 200 registros para fins de teste dada a limitação de recursos computacionais.

In [28]:
# df_reduced = df[df['Ano'] == 2015].copy()
df_reduced = df.head(200)
df_reduced.shape

(200, 16)

In [29]:
sentences = df_reduced.apply(generate_sentence, axis=1).to_list()

In [30]:
from transformers import AutoTokenizer, AutoModel
import torch

c:\Users\anycs\Documents\MATERIAL_11_PERIODO\PLN\Atividade 1\PLN_2024_Silva_Any\venv\Lib\site-packages\tqdm\auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


In [31]:
def text_to_embedding(sentences):
    # Load model from HuggingFace Hub
    tokenizer = AutoTokenizer.from_pretrained('rufimelo/Legal-BERTimbau-sts-base-ma-v2')
    model = AutoModel.from_pretrained('rufimelo/Legal-BERTimbau-sts-base-ma-v2')

    # Tokenize sentences
    encoded_input = tokenizer(sentences, padding=True, truncation=True, return_tensors='pt')

    # Compute token embeddings
    with torch.no_grad():
        model_output = model(**encoded_input)
    
    """ O primeiro elemento de model_output contém os embeddings dos tokens  """
    token_embeddings = model_output[0] #First element of model_output contains all token embeddings
    """ Expande a attention_mask (máscara que indica quais tokens são válidos e quais são padding) 
    para que tenha a mesma forma que token_embeddings. Isso é usado para mascarar os tokens de padding na média das embeddings.
    """
    input_mask_expanded = encoded_input['attention_mask'].unsqueeze(-1).expand(token_embeddings.size()).float()
    """ Calcula a média ponderada das embeddings dos tokens válidos, ignorando os tokens de padding. """
    return torch.sum(token_embeddings * input_mask_expanded, 1) / torch.clamp(input_mask_expanded.sum(1), min=1e-9)

In [32]:
sentence_embeddings = text_to_embedding(sentences)

Asking to truncate to max_length but no maximum length is provided and the model has no predefined maximum length. Default to no truncation.


In [33]:
type(sentence_embeddings)

torch.Tensor

In [34]:
sentence_embeddings.size()

torch.Size([200, 768])

## **Armazenamento de Embeddings no Milvus**

In [35]:
from pymilvus import connections, utility, MilvusException, \
    FieldSchema, CollectionSchema, DataType, Collection, Index

In [36]:
connections.connect(host="localhost", port="19530")

try:
    collections = utility.list_collections()
    print(f"Lista de coleções:\n", collections)
except MilvusException as e:
    print(e)

Lista de coleções:
 []


In [37]:
fields = [
    FieldSchema(name='id', dtype=DataType.INT64, is_primary=True),
    FieldSchema(name='embedding', dtype=DataType.FLOAT_VECTOR, dim=768)
]

In [38]:
schema = CollectionSchema(fields, "Amendments Collection")

In [39]:
collection = Collection('amendment_collection', schema)

In [40]:
index_params = {
    "index_type": "IVF_FLAT",
    "metric_type": "L2", 
    "params" : {"nlist": 100}
}

collection.create_index(field_name="embedding", index_params=index_params)

Status(code=0, message=)

In [41]:
embeddings_list = sentence_embeddings.tolist()

ids = list(range(len(embeddings_list)))


entities = [
    ids,              
    embeddings_list 
]

try:
    collection.insert(entities)
    print("Dados inseridos com sucesso!")
except Exception as e:
    print(f"Erro ao inserir dados: {e}")

Dados inseridos com sucesso!


In [42]:
collection.load()

In [43]:
try:
    # List all collections
    collections = utility.list_collections()
    print(f"List all collections:\n", collections)
except MilvusException as e:
    print(e)

List all collections:
 ['amendment_collection']


## **Busca por Similaridade Semântica**

### **Exemplo 1**

In [44]:
query_text = "2510659" # código IBGE
query_embedding = text_to_embedding(query_text).tolist()

Asking to truncate to max_length but no maximum length is provided and the model has no predefined maximum length. Default to no truncation.


In [45]:
# Ajustar parâmetros de pesquisa
search_params = {
    "metric_type": "L2",
    "params": {"nprobe": 10}
}

# Executar a pesquisa
results = collection.search(
    data=query_embedding,  # Use apenas o embedding da consulta
    anns_field="embedding",
    param=search_params,
    limit=10,
    consistency_level="Strong"
)

# Processar resultados
for result in results:
    print(f"IDs: {result.ids}, Distâncias: {result.distances}")

IDs: [154, 155, 45, 20, 25, 52, 77, 71, 112, 57], Distâncias: [101.50675964355469, 102.58380889892578, 109.434326171875, 115.203125, 115.32672119140625, 115.36993408203125, 117.16376495361328, 117.78767395019531, 118.3617172241211, 118.52607727050781]


In [46]:
sentences[154], sentences[155]

('Em fevereiro de 2015, o Município de São José dos Cordeiros - PB, com o código Siafi 2197 e o código IBGE 2514800, recebeu uma transferência no valor de R$ 125000. A Emenda Individual parlamentar classificada como DESPESAS DE CAPITAL teve como favorecido o CNPJ 8873226000109, identificado pelo nome MUNICIPIO DE SAO JOSE DOS CORDEIROS com a ordem bancária 135003000012015OB800208.',
 'Em fevereiro de 2015, o Município de Parari - PB, com o código Siafi 0494 e o código IBGE 2510659, recebeu uma transferência no valor de R$ 125000. A Emenda Individual parlamentar classificada como DESPESAS DE CAPITAL teve como favorecido o CNPJ 1612532000142, identificado pelo nome MUNICIPIO DE PARARI com a ordem bancária 135003000012015OB800209.')

### **Exemplo 2**

In [47]:
city = 'Pinhal Grande - RS'
selected_sentences = [s for s in sentences if city in s]
selected_sentences

['Em janeiro de 2015, o Município de Pinhal Grande - RS, com o código Siafi 5787 e o código IBGE 4314472, recebeu uma transferência no valor de R$ 97500. A Emenda Individual parlamentar classificada como DESPESAS DE CAPITAL teve como favorecido o CNPJ 94444346000122, identificado pelo nome MUNICIPIO DE PINHAL GRANDE com a ordem bancária 135098000012015OB800014.',
 'Em janeiro de 2015, o Município de Pinhal Grande - RS, com o código Siafi 5787 e o código IBGE 4314472, recebeu uma transferência no valor de R$ 97500. A Emenda Individual parlamentar classificada como DESPESAS DE CAPITAL teve como favorecido o CNPJ 94444346000122, identificado pelo nome MUNICIPIO DE PINHAL GRANDE com a ordem bancária 135098000012015OB800015.',
 'Em janeiro de 2015, o Município de Pinhal Grande - RS, com o código Siafi 5787 e o código IBGE 4314472, recebeu uma transferência no valor de R$ 243750. A Emenda Individual parlamentar classificada como DESPESAS DE CAPITAL teve como favorecido o CNPJ 94444346000122,

In [48]:
query_text = "Pinhal Grande - RS" # código IBGE
query_embedding = text_to_embedding(query_text).tolist()

Asking to truncate to max_length but no maximum length is provided and the model has no predefined maximum length. Default to no truncation.


In [49]:
# Ajustar parâmetros de pesquisa
search_params = {
    "metric_type": "L2",
    "params": {"nprobe": 10}
}

# Executar a pesquisa
results = collection.search(
    data=query_embedding,
    anns_field="embedding",
    param=search_params,
    limit=10,
    consistency_level="Strong"
)

# Processar resultados
for result in results:
    print(f"IDs: {result.ids}, Distâncias: {result.distances}")

IDs: [189, 31, 52, 20, 157, 12, 11, 17, 50, 112], Distâncias: [82.25318908691406, 84.67411041259766, 85.75894165039062, 87.2471923828125, 88.23783111572266, 89.07859802246094, 89.28851318359375, 90.22029113769531, 90.59690856933594, 91.22309875488281]


In [50]:
sentences[189], sentences[31]

('Em fevereiro de 2015, o Município de Pinhalzinho - SC, com o código Siafi 8253 e o código IBGE 4212908, recebeu uma transferência no valor de R$ 135160. A Emenda Individual parlamentar classificada como DESPESAS DE CAPITAL teve como favorecido o CNPJ 83021857000115, identificado pelo nome MUNICIPIO DE PINHALZINHO com a ordem bancária 135003000012015OB800243.',
 'Em janeiro de 2015, o Município de Pinhalzinho - SC, com o código Siafi 8253 e o código IBGE 4212908, recebeu uma transferência no valor de R$ 146250. A Emenda Individual parlamentar classificada como DESPESAS DE CAPITAL teve como favorecido o CNPJ 83021857000115, identificado pelo nome MUNICIPIO DE PINHALZINHO com a ordem bancária 135098000012015OB800037.')

> Isso acontece pois algoritmos de indexação aproximada, como IVF_FLAT, priorizam a eficiência da busca sobre a precisão absoluta. Eles podem retornar os vetores mais próximos, mas não garantem sempre a correspondência exata.

### **Exemplo 3**

In [51]:
city = 'Aracaju'
selected_sentences = [s for s in sentences if city in s]
selected_sentences

[]

In [52]:
query_text = "Aracaju"
query_embedding = text_to_embedding(query_text).tolist()

Asking to truncate to max_length but no maximum length is provided and the model has no predefined maximum length. Default to no truncation.


In [53]:
# Ajustar parâmetros de pesquisa
search_params = {
    "metric_type": "L2",
    "params": {"nprobe": 10}
}

# Executar a pesquisa
results = collection.search(
    data=query_embedding,
    anns_field="embedding",
    param=search_params,
    limit=10,
    consistency_level="Strong"
)

# Processar resultados
for result in results:
    print(f"IDs: {result.ids}, Distâncias: {result.distances}")

IDs: [112, 155, 72, 71, 45, 64, 133, 20, 77, 26], Distâncias: [92.51423645019531, 92.82566833496094, 93.04058074951172, 93.2039566040039, 93.73637390136719, 94.37173461914062, 95.0669937133789, 96.61090087890625, 96.61154174804688, 96.86224365234375]


In [54]:
sentences[112], sentences[155]

('Em fevereiro de 2015, o Município de Novo Itacolomi - PR, com o código Siafi 5517 e o código IBGE 4117297, recebeu uma transferência no valor de R$ 193000. A Emenda Individual parlamentar classificada como DESPESAS DE CAPITAL teve como favorecido o CNPJ 95639472000103, identificado pelo nome MUNICIPIO DE NOVO ITACOLOMI com a ordem bancária 135003000012015OB800166.',
 'Em fevereiro de 2015, o Município de Parari - PB, com o código Siafi 0494 e o código IBGE 2510659, recebeu uma transferência no valor de R$ 125000. A Emenda Individual parlamentar classificada como DESPESAS DE CAPITAL teve como favorecido o CNPJ 1612532000142, identificado pelo nome MUNICIPIO DE PARARI com a ordem bancária 135003000012015OB800209.')